Lambda functions get cards from real db and add cards to test db

MarkFrankle opened the pull request.

Commit:
Lambda functions get cards from real db and add cards to test db

ellistarn commented on nodeServerless/Local CE SAM/CE_lambda_functions/card_get.js:
```diff
@@ -0,0 +1,62 @@
+        var id = event.queryStringParameters.scryfallid;
+        var name = event.queryStringParameters.name;
+
+        console.log("id: " + id + ", type: " + typeof id + ". name: " + name + ", type: " + typeof name);
```
String interpolation
```
console.log(`id: ${id}, type: ${type}`)
```

ellistarn commented on nodeServerless/Local CE SAM/CE_lambda_functions/card_get.js:
```diff
@@ -0,0 +1,62 @@
+        console.log("id: " + id + ", type: " + typeof id + ". name: " + name + ", type: " + typeof name);
+        
+        var params = {
+            TableName: "BulkMagicCards",
```
This table name is weird. Perhaps just MagicCards?

MarkFrankle replied:
Can't change a table name in dynamodb. I'm planning on rewriting my code for bulk uploading in node instead of java, will switch to a new table then.

ellistarn replied:
This should be trivial to do if you tables are defined in CloudFormation. That's fine to punt. Why java?

MarkFrankle replied:
I was planning on doing the project in Java but a wise mentor taught me that strongly typed is a pita so I'm learning node instead.
Not sure what you mean by "This should be trivial to do if you tables are defined in CloudFormation"

ellistarn replied:
Set it up like this: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-dynamodb-table.html#aws-resource-dynamodb-table-syntax

Then you can just change the name and update your stack. You can even use the DDB table name  output as a Lambda environment variable to autowire your table name, but that may be a little too fancy ;)

MarkFrankle replied:
This is a cool way to do it. Looks like it only accepts string, number, boolean for data types. How should I deal with the lists in my json object?

ellistarn replied:
There is absolutely a list type in DDB. If you're in node, feel free to go schemaless.

ellistarn commented on nodeServerless/Local CE SAM/CE_lambda_functions/card_get.js:
```diff
@@ -0,0 +1,62 @@
+        var params = {
+            TableName: "BulkMagicCards",
+            Key: {
+                'scryfallid' : id,
```
what's this?

MarkFrankle replied:
What's what? params is the search parameters, tablename is self-explanatory, scryfallid is the primary key for my db because it's the internal id from the api I'm using for cards, a service called scryfall. Will switch to kabob-case

ellistarn replied:
I was talking specifically about the scryfallid. Mostly out of curiosity, the code looks fine for this. Is that your data source?

MarkFrankle replied:
Yep. https://scryfall.com/docs/api

ellistarn commented on nodeServerless/Local CE SAM/CE_lambda_functions/card_get.js:
```diff
@@ -0,0 +1,62 @@
+// let ddb = new AWS.DynamoDB();
+
+exports.lambdaHandler = async (event, context) => {
+    try {
```
Probably a bad idea to stuff everything in a big try/catch. Use try/catch intentionally and with precision.

ellistarn commented on nodeServerless/Local CE SAM/CE_lambda_functions/card_get.js:
```diff
@@ -0,0 +1,62 @@
+    
+};
+
+// return response
```
Bad practice to leave commented code.

ellistarn commented on nodeServerless/Local CE SAM/CE_lambda_functions/card_get.js:
```diff
@@ -0,0 +1,62 @@
+//         message: id,
+//         // location: ret.data.trim()
+//     })
+// }
```
newline at the end of file helps with readability.

ellistarn commented on nodeServerless/Local CE SAM/CE_lambda_functions/card_get.js:
```diff
@@ -0,0 +1,62 @@
+        
+        var params = {
+            TableName: "BulkMagicCards",
+            Key: {
```
Why are you using name and id as a key? Surely id is unique enough?

MarkFrankle replied:
I'm using id as the primary key and name as the sort key, and both keys are required by the DocumentClient to get. Should I not use a sort key here?

ellistarn commented on nodeServerless/Local CE SAM/CE_lambda_functions/card_get.js:
```diff
@@ -0,0 +1,62 @@
+let docClient = new AWS.DynamoDB.DocumentClient();
+// let ddb = new AWS.DynamoDB();
+
+exports.lambdaHandler = async (event, context) => {
```
It's up to you, but I prefer promises in this case. try/catch is kind of a code ick. 

It might be cleaner to write this of the form:

```
exports.lambdaHandler = async (event, context) => {
if (!validate(event)) {
  return {
    statusCode: 400,
    error: `Expected input of the form...`
  };
}

return docClient.get({
  // put the whole payload inline
}).promise().then((data) => {
  return { statusCode: 200, body: JSON.stringify(data) };
}).catch(error) => {
  return {
    statusCode: 500,
    error: `Could not fetch: ${error.stack}`
  };
});

```

MarkFrankle replied:
By "put the whole payload inline" I assume you mean define my params var there?
```
return docClient.get({
     var params = {
        TableName: "BulkMagicCards",
        Key: {
            'scryfallid' : id,
            'name' : name
        }      
    };
}).promise().then...
```

ellistarn replied:
Yeah, almost.

```
return docClient.get({
    TableName: "BulkMagicCards",
    Key: {
        'scryfallid' : id,
        'name' : name
    }      
}).promise().then...

```

ellistarn commented on nodeServerless/Local CE SAM/CE_lambda_functions/card_get.js:
```diff
@@ -0,0 +1,62 @@
+            return { statusCode: 200, body: JSON.stringify(data) };
+        } catch (error) {
+            return {
+              statusCode: 400,
```
4xx is user error, 5xx is server error.
https://en.wikipedia.org/wiki/List_of_HTTP_status_codes

ellistarn commented on nodeServerless/Local CE SAM/CE_lambda_functions/card_get.js:
```diff
@@ -0,0 +1,62 @@
+})
+
+let docClient = new AWS.DynamoDB.DocumentClient();
+// let ddb = new AWS.DynamoDB();
```
delete

ellistarn commented on nodeServerless/Local CE SAM/CE_lambda_functions/card_get.js:
```diff
@@ -0,0 +1,62 @@
+    region: "us-west-2"
+})
+
+let docClient = new AWS.DynamoDB.DocumentClient();
```
consider just `ddb`

MarkFrankle replied:
There's a different object for managing this that is a DynamoDB object, so I prefer this for clarity.

ellistarn commented on nodeServerless/Local CE SAM/CE_lambda_functions/card_get.js:
```diff
@@ -0,0 +1,62 @@
+exports.lambdaHandler = async (event, context) => {
+    try {
+        console.log("\n\n*********************\n");
+        var id = event.queryStringParameters.scryfallid;
```
Maybe im missing some context, but query parameters are the wrong way to encode these. These should probably be encoded in a request body or uri path.

Query strings are for optional flags like query filters. Webapps and automatic webcaching systems treat these differently which is why it's important to follow the convention.

MarkFrankle replied:
I thought that search parameters did go in the query string? I'm looking at this stackoverflow right now https://stackoverflow.com/a/45462793. What's the convention?

ellistarn commented on nodeServerless/Local CE SAM/CE_lambda_functions/card_put.js:
```diff
@@ -0,0 +1,70 @@
+
+let response;
+var AWS = require('aws-sdk');
+AWS.config.update({
```
It would be nice if you could configure this in one place for all your handlers. Can you set this as an environment variable in your SAM function?

MarkFrankle replied:
This assumes that something is running first, before the lambda function, that would initialize this. Idk about polymorphism in JS but maybe a parent class that deals with it? Not sure where I'd initialize this besides where it is now
(I found this https://github.com/awslabs/serverless-application-model/blob/master/docs/globals.rst and don't think it would do what you're looking for?)

MarkFrankle replied:
Looking into logging and it looks like I have to create and configure my logger in every file as well?

ellistarn commented on nodeServerless/Local CE SAM/template.yaml:
```diff
@@ -0,0 +1,37 @@
+        HelloWorld:
+          Type: Api # More info about API Event Source: https://github.com/awslabs/serverless-application-model/blob/master/versions/2016-10-31.md#api
+          Properties:
+            Path: /card
```
resources should be plural in uri.

POST hostname/myresource/plural.

ellistarn commented on nodeServerless/Local CE SAM/template.yaml:
```diff
@@ -0,0 +1,37 @@
+    Type: AWS::Serverless::Function # More info about Function Resource: https://github.com/awslabs/serverless-application-model/blob/master/versions/2016-10-31.md#awsserverlessfunction
+    Properties:
+      CodeUri: CE_lambda_functions/
+      Handler: card_get.lambdaHandler
```
snake_case is very python. JS is typically camelCase or kabob-case

ellistarn commented on nodeServerless/Local CE SAM/event.json:
```diff
@@ -0,0 +1,63 @@
+  "path": "/path/to/resource",
+  "httpMethod": "GET",
+  "isBase64Encoded": false,
+  "queryStringParameters": {
```
make these path params

MarkFrankle replied:
I need the hash key and the sort key to call the documentclient get operation. Should my path be "/cards/cardname-cardID/" then? Should I use a sort key generally?

ellistarn commented on nodeServerless/Local CE SAM/template.yaml:
```diff
@@ -0,0 +1,37 @@
+# More info about Globals: https://github.com/awslabs/serverless-application-model/blob/master/docs/globals.rst
+Globals:
+  Function:
+    Timeout: 3
```
3 seconds? 3 minutes?

ellistarn commented on nodeServerless/Local CE SAM/template.yaml:
```diff
@@ -0,0 +1,37 @@
+  # ServerlessRestApi is an implicit API created out of Events key under Serverless::Function
+  # Find out more about other implicit resources you can reference within SAM
+  # https://github.com/awslabs/serverless-application-model/blob/master/docs/internals/generated_resources.rst#api
+  HelloWorldApi:
```
I dont think you have a helloworld api. This doesn't matter that much, but other cloudformation stacks may want to consume these values -- which are currently somewhat bogus. Consider removing this section until that's necessary.

ellistarn reviewed (commented):
Your filenames are weird. Node recommends kabob-case for folder names. Never put whitespace in paths -- it breaks lots of things and is difficult to type.

Consider:

my-project/api/card/get.js
my-project/api/card/list.js
my-project/api/card/update.js
my-project/api/card/delete.js

ellistarn commented:
Consider using winston (https://github.com/winstonjs/winston) for logging

ellistarn commented:
Your commit messages should be in the imperative. Read https://chris.beams.io/posts/git-commit/

ellistarn commented on nodeServerless/Local CE SAM/CE_lambda_functions/card_put.js:
```diff
@@ -0,0 +1,70 @@
+            }       
+        };
+
+        let putItem = new Promise((res, rej) => {
```
consider using full words when reasonable (resolve, reject)

ellistarn replied:
Consider using the promise pattern I mentioned in the other API

ellistarn commented on nodeServerless/Local CE SAM/CE_lambda_functions/card_put.js:
```diff
@@ -0,0 +1,70 @@
+        };
+
+        let putItem = new Promise((res, rej) => {
+            docClient.put(params, function(err, data) {
```
HTTP Verb standards:
POST: Create a thing (not idempotent)
PUT: Replace a thing (idempotent)
PATCH: Merge a thing w/ what exists (idempotent)
DELETE: Delete a thing (idempotent or not, depends on callers)

ellistarn commented:
Nice work overall. I know how difficult it is to learn async paradigms and integrate with systems like lambda. I've made a bunch of best practices comments. Typically it's good to:

1. Know file/variable naming/casing conventions for your language
2. Don't litter your code with comments. Don't leave dead code.
3. Console.log should never be checked in, use a logger.

ellistarn commented on nodeServerless/Local CE SAM/CE_lambda_functions/card_get.js:
```diff
@@ -0,0 +1,62 @@
+ * 
+ */
+
+let response;
```
toss this global let

Commit:
Remove excess autogen code and fix best practices

 - Reformat directory structure to match api format
   - Rename files to match api definition
 - Trim template.yaml to match project as it exists
 - Cleaned code in get and put functions

Commit:
Standardize promise pattern and rename put to post

Commit:
Remove console log

Commit:
Adopt AirBnB style guidelines

Commit:
Fix string interpolation

ellistarn commented:
I can't find the thread on querystrings.

Typically your resources should be fully identifyable at a resource path:
GET: `/cards/id`

You may also want to write some queryable endpoints like
GET: `/cards?year=2019&minPrice=25` Returns the search result
GET: `/cards` Returns all cards (table scan or page)

It's typically bad practice to use data on GET requests and bad practice to use querystring on PUT/POST/PATCH/DELETE. Data is used for POST/PATCH to represent the entire object you're "posting"/"patching". 

For POST/PATCH/PUT, you typically do
POST: `/cards` { obj: ... }, ID is extracted from obj.id, typically.
PUT: `/cards/id` { obj: ...}, Replaces object
PATCH: `/cards/id` { patch: ... }, [Updates object](https://en.wikipedia.org/wiki/JSON_Patch)
